### PR TITLE
[Libomptarget] Remove __tgt_image_info and use the ELF directly

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
This patch reorganizes a lot of the code used to check for compatibility
with the current environment. The main bulk of this patch involves
moving from using a separate `__tgt_image_info` struct (which just
contains a string for the architecture) to instead simply checking this
information from the ELF directly. Checking information in the ELF is
very inexpensive as creating an ELF file is simply writing a base
pointer.

The main desire to do this was to reorganize everything into the ELF
image. We can then do the majority of these checks without first
initializing the plugin. A future patch will move the first ELF checks
to happen without initializing the plugin so we no longer need to
initialize and plugins that don't have needed images.

This patch also adds a lot more sanity checks for whether or not the ELF
is actually compatible. Such as if the images have a valid ABI, 64-bit
width, executable, etc.

### CODE REVIEW

**Comment by ronlieb** [comment]
Thanks for patch , Saiislam is actively reviewing over the next 2 days.

File: openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp [hunk 1]
```diff
@@ -26,52 +26,228 @@ using namespace llvm;
...
+    return Flags & EF_CUDA_ACCELERATORS ? "sm_90a" : "sm_90";
+  default:
+    return createError("Unknown CUDA architecture");
+  }
+}
 
-  if (!Object) {
-    DP("Unknown ELF format!\n");
-    return 0;
+static Expected<StringRef> getAMDGPUProcessor(uint32_t Flags) {
```
Comment (jplehr) [inline openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:125]: Not really part of this patch, but should this generally live somewhere to be reused by more than just libomptarget?

Comment (jhuber6) [reply openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:125]: I could put it in `~/Documents/llvm/llvm-project/llvm/lib/BinaryFormat/ELF.cpp` next to `convertEMachineToArchName` if needed.

**Comment by jhuber6** [comment]
> This patch doesn't seem to work with `clang -O2 -fopenmp --offload-arch=gfx90a:sramecc+:xnack- --offload-arch=gfx90a:sramecc+:xnack+ helloworld.c -o helloworld`
> 
> HSA_XNACK=1 ./helloworld HSA_XNACK=0 ./helloworld
> 
> Both of them should work on a system which allows HSA_XNACK toggling on per process basis.
> 
> I have couple of more concerns for which I am trying to get counter cases.

So, the desired behavior here is that the prior should reject `xnack-` while the latter should reject `xnack+`. I think in the logic I only check if the opposite is set for returning an invalid image, but it seems I need to check that the target ID contains the equivalent value.

File: openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h [hunk 2]
```diff
@@ -58,92 +59,58 @@ uint32_t getImplicitArgsSize(uint16_t Version) {
...
-  }
 
-  // Incompatible if image has more features than the environment,
-  // irrespective of type or sign of features.
-  if (ImgMap.size() > EnvMap.size()) {
-    DP("Incompatible: Image has more features than the Environment \t[Image: "
-       "%s]\t:\t[Env: %s]\n",
-       ImageTargetID.data(), EnvTargetID.data());
-    return false;
+  // Check if the image is requesting xnack on or off.
```
Comment (saiislam) [inline openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:79]: Why not use `amdhsa.target` string from the metadata and pass it is as is to the existing function?

https://llvm.org/docs/AMDGPUUsage.html#amdgpu-amdhsa-code-object-metadata-map-table-v4

Comment (jhuber6) [reply openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:79]: Does that net us anything? The information seems freely available in the ELF flags which is easier to retrieve than the AMDHSA metadata that's stored in a YAML form if memory serves. This way we don't need to parse anything and it's fast in the common case (No special flags set).

**Review by doru1004 — COMMENTED** [review]
Can we have tests for this patch?

**Comment by jhuber6** [comment]
> Can we have tests for this patch?

The existing tests not breaking should be good enough. Testing the target-id stuff directly is difficult because it would depend on the platform you're running on.

**Comment by doru1004** [comment]
> > Can we have tests for this patch?
> 
> The existing tests not breaking should be good enough. Testing the target-id stuff directly is difficult because it would depend on the platform you're running on.

Could we add one of those CHECK tests which looks at the output for AMD GPUs perhaps and make sure it emits what we expect? Not sure if that is feasible, just a thought.

**Comment by jhuber6** [comment]
> > > Can we have tests for this patch?
> > 
> > 
> > The existing tests not breaking should be good enough. Testing the target-id stuff directly is difficult because it would depend on the platform you're running on.
> 
> Could we add one of those CHECK tests which looks at the output for AMD GPUs perhaps and make sure it emits what we expect? Not sure if that is feasible, just a thought.

Might be doable with unit tests in the future. We can just export more of this stuff as a single library.

File: openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp [hunk 3]
```diff
@@ -1651,44 +1671,26 @@ int32_t __tgt_rtl_init_plugin() {
 }
 
 int32_t __tgt_rtl_is_valid_binary(__tgt_device_image *TgtImage) {
+  // TODO: We should be able to perform a trivial ELF machine check without
+  // initializing the plugin first to save time if the plugin is not needed.
```
Comment (jdoerfert) [inline openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp:1675]: Right, we need a "static" part and a post-init part. The former should weed out clear mismatches and the latter can check things like version numbers and flag support.
Follow up though.

File: openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h (hunk 2 above)
Comment (jdoerfert) [reply openmp/libomptarget/plugins-nextgen/amdgpu/utils/UtilitiesRTL.h:79]: If we get the information from the ELF, it makes sense to use it. Unsure why it's duplicated though.

**Review by jdoerfert — APPROVED** [review]
LG from me, I'll let AMD finish their testing.

File: openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp (hunk 1 above)
Comment (saiislam) [reply openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:125]: Can `getAMDGPUCPUName` from `llvm/lib/Object/ELFObjectFile.cpp` be used here instead of redefining this?

Otherwise, we will have to maintain this copy in sync with the original one at ELFObjectFile.cpp.

File: openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp [hunk 4]
```diff
@@ -26,52 +26,228 @@ using namespace llvm;
...
+  if (Header.e_type != ET_EXEC && Header.e_type != ET_DYN)
+    return createError("Only executable ELF files are supported");
+
+  if (Header.e_machine == EM_AMDGPU) {
+    if (Header.e_ident[EI_OSABI] != ELFOSABI_AMDGPU_HSA)
+      return createError("Invalid AMD OS/ABI, must be AMDGPU_HSA");
+    if (Header.e_ident[EI_ABIVERSION] != ELFABIVERSION_AMDGPU_HSA_V4 &&
+        Header.e_ident[EI_ABIVERSION] != ELFABIVERSION_AMDGPU_HSA_V5)
+      return createError("Invalid AMD ABI version, must be version 4 or 5");
+    if ((Header.e_flags & EF_AMDGPU_MACH) < EF_AMDGPU_MACH_AMDGCN_GFX700 ||
```
Comment (saiislam) [inline openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:64]: Better to leave this check to the clang driver, backend, or hsa runtime. All archs between 700 and 1201 are not supported and it doesn't seem worth to check it again here.

File: openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp (hunk 1 above)
Comment (jhuber6) [reply openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:125]: Didn't realize that was there, but it's a private method. I can probably modify that to work here and add one for NVPTX instead.

File: openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp (hunk 4 above)
Comment (jhuber6) [reply openmp/libomptarget/plugins-nextgen/common/src/Utils/ELF.cpp:64]: I mostly just wanted this as a sanity check for if someone ends up running something on a newer machine than we've accounted for. In the future the runtime is going to encompass more than what we support with OpenMP (like when I get around to writing a loader utilitly with the runtime like in `libc` ) So I wanted some basic checks beyond just the machine.

**Comment by jhuber6** [comment]
Changed the patch a little, realized there was no need for the actual SM architecture name since we can just check the minor and major values with the flags directly. Got rid of the indirection around getting the platform flags and just use the ELF interface directly.
